src/FundSourceApprover/FundSourceAllower.sol

```diff
@@ -0,0 +1,89 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.16;
+
+import "openzeppelin-contracts/contracts/token/ERC20/utils/SafeERC20.sol";
+
+contract FundSourceAllower {
+    using SafeERC20 for IERC20;
+
+    // chain id of chain in which funds initiate; used only for bookkeeping
+    uint256 public immutable sourceChaindId;
+    // address which eth is forwarded to
+    address public immutable ethDestination;
+    // address which tokens are forwarded to
+    address public immutable tokenDestination;
+    // address with affordance to approve funds
+    address public immutable admin;
+    // signifies whether admin has approved transfering of funds to destination
+    bool public approved;
+
+    error TransferFailed();
+    error NotApproved();
+    error NotFromAdmin(address sender);
+
+    event FundsTransfered(uint256 amount);
+    event ApprovedStateSet(bool approved);
+
+    constructor(uint256 _sourceChaindId, address _ethDestination, address _tokenDestination, address _admin) {
+        sourceChaindId = _sourceChaindId;
+        ethDestination = _ethDestination;
+        tokenDestination = _tokenDestination;
+        admin = _admin;
+    }
```

Comment:
could make admin msg.sender also

```suggestion
    constructor(uint256 _sourceChaindId, address _ethDestination, address _tokenDestination) {
        sourceChaindId = _sourceChaindId;
        ethDestination = _ethDestination;
        tokenDestination = _tokenDestination;
        admin = msg.sender;
    }
```

Reply:
i think i prefer this for testing / isolating the rationale of the components, if that makes sense